nalgebra-lapack/src/colpiv_qr.rs

```diff
@@ -0,0 +1,797 @@
+                check_lapack_info(info)?;
+                Ok(work[0] as i32)
+            }
+
+            unsafe fn xtrtrs(
+                uplo: TriangularStructure,
+                trans: Transposition,
+                diag: DiagonalKind,
+                n: i32,
+                nrhs: i32,
+                a: &[Self],
+                lda: i32,
+                b: &mut [Self],
+                ldb: i32,
+            ) -> Result<(), LapackErrorCode> {
+                let mut info = 0;
+                let trans = match trans {
+                    Transposition::No => b'N',
+                    Transposition::Transpose => b'T',
+                };
+
+                unsafe {
+                    $xtrtrs(
+                        uplo.into_lapack_uplo_character(),
+                        trans,
+                        diag.into_lapack_diag_character(),
+                        n,
+                        nrhs,
+                        a,
+                        lda,
+                        b,
+                        ldb,
+                        &mut info,
+                    );
+                }
+
+                check_lapack_info(info)
+            }
+
+            unsafe fn xlapmt(
+                forwrd: bool,
+                m: i32,
+                n: i32,
+                x: &mut [Self],
+                ldx: i32,
+                k: &mut [i32],
+            ) -> Result<(), LapackErrorCode> {
+                debug_assert_eq!(k.len(), n as usize);
+
+                let forward: [i32; 1] = [forwrd.then_some(1).unwrap_or(0)];
+                unsafe { $xlapmt(forward.as_slice(), m, n, x, ldx, k) }
+                Ok(())
+            }
+
+            unsafe fn xlapmr(
+                forwrd: bool,
+                m: i32,
+                n: i32,
+                x: &mut [Self],
+                ldx: i32,
+                k: &mut [i32],
+            ) -> Result<(), LapackErrorCode> {
+                debug_assert_eq!(k.len(), m as usize);
+
+                let forward: [i32; 1] = [forwrd.then_some(1).unwrap_or(0)];
+                unsafe { $xlapmr(forward.as_slice(), m, n, x, ldx, k) }
+                Ok(())
+            }
+        }
+    };
+}
+
+colpiv_qr_scalar_impl!(
+    f32,
+    xgeqp3 = lapack::sgeqp3,
+    xtrtrs = lapack::strtrs,
+    xlapmt = lapack::slapmt,
+    xlapmr = lapack::slapmr
+);
+
+colpiv_qr_scalar_impl!(
+    f64,
+    xgeqp3 = lapack::dgeqp3,
+    xtrtrs = lapack::dtrtrs,
+    xlapmt = lapack::dlapmt,
+    xlapmr = lapack::dlapmr
+);
+
+/// Trait implemented by reals for which Lapack function exist to compute the
+/// column-pivoted QR decomposition.
+// @note(geo-ant) This mirrors the behavior in the existing QR implementation
+// without pivoting. I'm not 100% sure that we can't abstract over real and
+// complex behavior in the scalar trait, but I'll keep it like this for now.
+#[allow(missing_docs)]
+pub unsafe trait ColPivQrReal: ColPivQrScalar + QRReal {
```

Comment:
I think that `unsafe` is not required here for a whole trait as merely implementing this trait does not make any safe code that does not use any `unsafe` methods of this trait itself unsafe.

Otherwise safety-related changes LGTM.